chain/client/src/stateless_validation/chunk_validator/orphan_witness_handling.rs

```diff
@@ -0,0 +1,165 @@
+//! This module contains the logic of handling orphaned chunk state witnesses.
+//! To process a ChunkStateWitness we need its previous block, but sometimes
+//! the witness shows up before the previous block is available, so it can't be
+//! processed immediately. In such cases the witness becomes an orphaned witness
+//! and it's kept in the pool until the required block arrives. Once the block
+//! arrives, all witnesses that were waiting for it can be processed.
+
+use crate::Client;
+use itertools::Itertools;
+use near_chain_primitives::Error;
+use near_primitives::stateless_validation::ChunkStateWitness;
+use near_primitives::types::{BlockHeight, EpochId};
+use std::ops::Range;
+
+/// We keep only orphan witnesses that are within this distance of
+/// the current chain head. This helps to reduce the size of
+/// OrphanStateWitnessPool and protects against spam attacks.
+/// The range starts at 2 because a witness at height of head+1 would
+/// have the previous block available (the chain head), so it wouldn't be an orphan.
+pub const ALLOWED_ORPHAN_WITNESS_DISTANCE_FROM_HEAD: Range<BlockHeight> = 2..6;
+
+/// We keep only orphan witnesses which are smaller than this size.
+/// This limits the maximum memory usage of OrphanStateWitnessPool.
+pub const MAX_ORPHAN_WITNESS_SIZE: usize = 40_000_000;
```

Comment:
> Perhaps add a TODO(stateless_validation) to merge those limits when the latter comes into effect?>

👍 , added a TODO(#10259)

> nit: use bytesize for sizes

Nice, I didn't know about it, added

> mini nit: Maybe make it configurable in case the witnesses turn out to be larger than we expect? Totally optional as long as the value has a safe margin over the expected max state witness size.

Yeah that could be nice, I didn't add it yet because I wanted to get the PR out there, but I'll think about it.